### PR TITLE
Add solution to week 5's project

### PR DESCRIPTION
**Pull request by RossJ24**

### CODE REVIEW

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 1]
```diff
@@ -22,6 +22,12 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
-    return "Hello " + name;
+    String cleanedName = "";
+    char [] nameRay = name.trim().toCharArray();
+    for(char c : nameRay){
+        cleanedName += (Character.isLetter(c) || Character.isDigit(c))? c : "";
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: I believe there is a Character.isAlphaNumeric?

Also, make sure to space the operators. 

Finally, Use a StringBuilder for cleanedName

if (...) {
  cleanedName.append(c);
}

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 2]
```diff
@@ -22,6 +22,12 @@
...
    */
   public String greet(String name) {
-    return "Hello " + name;
+    String cleanedName = "";
+    char [] nameRay = name.trim().toCharArray();
+    for(char c : nameRay){
+        cleanedName += (Character.isLetter(c) || Character.isDigit(c))? c : ""; 
+    }
+    
+    return "Hello " + cleanedName;
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: Move "Hello " to cleanedname initializer

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 3]
```diff
@@ -30,4 +30,22 @@ public void testGreeting() {
...
+	}
+	@Test
+	public void testGreetingOddCharsWithWhitespace() {
+	Greeter greeter = new Greeter();
+
+	String greeting = greeter.greet("   Ross$$$");
+
+	// Whitespace should be trimmed
+	Assert.assertEquals("Hello Ross", greeting);
+	}
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: Formatting is off.

Testing is broken into three parts:

Variable instantiation

Execution of unit test

Assertion data is correct.

In last two steps, do not inline testing data, keep it with the 1st section and call upon the variable.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,158 @@
...
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> ret = new ArrayList<TimeRange>();
+    if(request.getDuration() > 1440){
+        return ret;
+    }
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: move this above variable declaration above. Then add a newline between this and the next check.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,158 @@
...
+    if(events.size() == 0) {
+        ret.add(TimeRange.WHOLE_DAY);
+        return ret;
+    }
+    ArrayList<Event> otherMeetings;
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:37]: new line between method bounds section

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,158 @@
...
+    if(request.getOptionalAttendees().size() >= 0 && request.getAttendees().size() == 0)
+        otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, true);
+    else{
+        otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, false);
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: inline if statement as it's a true / false statement

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,158 @@
...
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> ret = new ArrayList<TimeRange>();
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: please change name here to be more indicative of what it's purpose is.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,158 @@
...
+    if(request.getOptionalAttendees().size() >= 0 && request.getAttendees().size() == 0)
+        otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, true);
+    else{
+        otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, false);
+    }
+    if(otherMeetings.size() == 0) {
+        ret.add(TimeRange.WHOLE_DAY);
+        return ret;
+    }
+        ArrayList<TimeRange> timeranges =  new ArrayList<TimeRange>();
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: timeRanges

Also ArrayList<TimeRange> timeRanges = otherMeetings.stream().map(Event::getWhen).collect(Collectors.asList());

I think there is a sort function as well for streams?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,158 @@
...
+        for(Event e : otherMeetings){
+            timeranges.add(e.getWhen());
+        }
+        Collections.sort(timeranges, TimeRange.ORDER_BY_START);
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: delete based on above comment.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 10]
```diff
@@ -270,5 +271,107 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void optionalAttendeeIsNotConsidered() {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:276]: follow test principles outlined above in other test file.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -15,9 +15,158 @@
...
+    * @return new TimeRange created from the overlap between tr1 and tr2
+   */
+  public TimeRange mergeTimeRanges(TimeRange tr1, TimeRange tr2) {
+      int earliest;
+      int latest;
+      if(tr1.start() < tr2.start()) {
+        earliest = tr1.start();
+      } else {
+        earliest = tr2.start();
+      }
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: earliest = Math.min(tr1.start(), tr2.start());

same for latest.

Also, where is the check to ensure that the timelines do overlap?

**Review by KillianLeC — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java (hunk 3 above)
Comment (RossJ24) [reply walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: I was just following the format they gave

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (RossJ24) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: That is taken care of in the previous function

File: walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java [hunk 12]
```diff
@@ -22,6 +22,12 @@
    * Returns a greeting for the given name.
    */
   public String greet(String name) {
-    return "Hello " + name;
+    StringBuilder cleanedName = new StringBuilder("Hello ");
+    char [] nameRay = name.trim().toCharArray();
+    for(char c : nameRay){
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/main/java/com/google/sps/Greeter.java]: add the char[] directly to the for loop. 

for (char c : name.trim().toCharArray())

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 13]
```diff
@@ -30,4 +30,22 @@ public void testGreeting() {
...
+	@Test
+	public void testGreetingOddCharsWithWhitespace() {
+	Greeter greeter = new Greeter();
+
+	String greeting = greeter.greet("   Ross$$$");
+
+	// Whitespace should be trimmed and special characters taken out
+	Assert.assertEquals("Hello Ross", greeting);
+	}
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: indentation

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 14]
```diff
@@ -30,4 +30,22 @@ public void testGreeting() {
 
     Assert.assertEquals("Hello Ada", greeting);
   }
+	@Test
+	public void testGreetingTrimsWhitespace() {
+	Greeter greeter = new Greeter();
+
+	String greeting = greeter.greet("   Ada   ");
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java]: Greeter greeter = ...
String name = "      Ada    ";
String result = "Hello Ada";

Do same for other test case.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -15,9 +15,152 @@
...
+    }
+
+    ArrayList<Event> otherMeetings;
+    if(request.getOptionalAttendees().size() >= 0 && request.getAttendees().size() == 0) {
+         otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, true);
+    } 
+    else {
+        otherMeetings = doAttendeesHaveAnyOtherMeetings(events, request, false);
+    }
+    if(otherMeetings.size() == 0) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: otherMeetings.isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -15,9 +15,152 @@
...
+        return ret;
+    }
+
+    if(events.size() == 0) {
+        ret.add(TimeRange.WHOLE_DAY);
+        return ret;
+    }
+
+    ArrayList<Event> otherMeetings;
+    if(request.getOptionalAttendees().size() >= 0 && request.getAttendees().size() == 0) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: !request.getOptionalAttendees().isEmpty()

request.getAttendees.isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -15,9 +15,152 @@
...
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Return value of the query
+    ArrayList<TimeRange> ret = new ArrayList<TimeRange>();
+    if(request.getDuration() > 1440){
+        return ret;
+    }
+
+    if(events.size() == 0) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: events.isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -15,9 +15,152 @@
...
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Return value of the query
+    ArrayList<TimeRange> ret = new ArrayList<TimeRange>();
+    if(request.getDuration() > 1440){
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: please move 1440 to a class global field indicating what that timespan means.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -15,9 +15,152 @@
...
+            if(newTimeRange.start() != newTimeRange.end() && newTimeRange.duration() >= request.getDuration()){
+                ret.add(newTimeRange);
+            }
+            currentStart =  tr.end();
+        }
+        TimeRange newTimeRange = TimeRange.fromStartEnd(currentStart,1440,false);
+        if(newTimeRange.start() != newTimeRange.end() && newTimeRange.duration() >= request.getDuration()) {
+            ret.add(newTimeRange);
+        } 
+        if(request.getOptionalAttendees().size() > 0) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: !request.getOptionalAttendees().isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -15,9 +15,152 @@
...
+  }
+
+  /**
+    * Function that determines whether any attendees of the requested meeting have any other meeting that day
+    * @param events a collection of events
+    * @param request a meeting request 
+    * @param optional a boolean flag that determines whether the function checks optional attendees or mandatory attendees
+    * @return ret an ARraylist of the other Events
+    */
+  public ArrayList<Event> doAttendeesHaveAnyOtherMeetings(Collection<Event> events, MeetingRequest request, boolean optional) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: rename function. The current name indicates the function returns a boolean. You want to return all meetings for the day.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -15,9 +15,152 @@
...
+      ArrayList<Event> ret = new ArrayList<Event>();
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: return events.stream().filter((e) -> doAttendeesOverlap(e, request, optional)).collect(Collectors.toList());

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -15,9 +15,152 @@
...
+            ret.add(e);
+        }
+      }
+      return ret;
+  }
+
+  /**
+    * Function that checks a single event and determines whether the requested meetings attendees are included in it
+    * @param event the Event that the MeetingRequest will be checked againt
+    * @param request the requested Meeting
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:86]: missing one parameter.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -15,9 +15,152 @@
...
+            ret = tryToFitOptionalAttendees(ret,events,request);
+        }
+        return ret;
+  }
+
+  /**
+    * Function that determines whether any attendees of the requested meeting have any other meeting that day
+    * @param events a collection of events
+    * @param request a meeting request 
+    * @param optional a boolean flag that determines whether the function checks optional attendees or mandatory attendees
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:76]: is optional flag indicating to include both optional attendees and mandatory attendees? Or is it just optional attendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -15,9 +15,152 @@
...
+  public boolean doAttendeesOverlap(Event event, MeetingRequest request, boolean optional) {
+      for(String eventAttendee : event.getAttendees()) {
+          for(String requestAttendee : optional ? request.getOptionalAttendees() : request.getAttendees()) {
+              if(eventAttendee.compareTo(requestAttendee) == 0) {
+                  return true;
+              }
+          }
+      }
+      return false;
+  }
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This function seems like a misnomer? It checks if there are attendees from one event to a MeetingRequest?

Maybe isOverlapBetweenAttendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 25]
```diff
@@ -15,9 +15,152 @@
...
+  /**
+    * Function that iterates through the timeranges and removes and overlapping timeranges
+    * @param timeranges an Arraylist of TimeRanges
+    * @return void 
+    */
+  public void removeOverlap(ArrayList<TimeRange> timeranges){
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: performUnionOverTimeRanges?

Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The time ranges need to be sorted, please sort in this function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 26]
```diff
@@ -15,9 +15,152 @@
...
+    * Function that iterates through the timeranges and removes and overlapping timeranges
+    * @param timeranges an Arraylist of TimeRanges
+    * @return void 
+    */
+  public void removeOverlap(ArrayList<TimeRange> timeranges){ 
+      for(int i = 1; i < timeranges.size(); ++i){
+        if(timeranges.get(i-1).overlaps(timeranges.get(i))) {
+            TimeRange newrange = mergeTimeRanges(timeranges.get(i-1), timeranges.get(i));
+            timeranges.set(i-1,newrange);
+            timeranges.remove(i);
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:110]: i-- after a remove

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 27]
```diff
@@ -15,9 +15,152 @@
...
+        }
+      }
+  }
+  /**
+    * Function that merges two overalapping timeranges
+    * @param tr1 a TimeRange that needs to be merged
+    * @param tr2 another TimeRange that needs to be merged
+    * @return new TimeRange created from the overlap between tr1 and tr2
+   */
+  public TimeRange mergeTimeRanges(TimeRange tr1, TimeRange tr2) {
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: private function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 28]
```diff
@@ -15,9 +15,152 @@
...
+    * Function that iterates through each event, looking for one that conflicts with optional attendees of the request
+    * @param timeranges an arraylist of timeranges that is to be reduced
+    * @param events a collection of events that need to be checked against
+    * @param request the meeting request
+    * @return an arraylist of timeranges with the conflicting ranges removed
+    */
+  public ArrayList<TimeRange> tryToFitOptionalAttendees(ArrayList<TimeRange> timeranges, Collection<Event> events, MeetingRequest request) {
+    ArrayList<String> optionalAttendees = new ArrayList<String>(request.getOptionalAttendees());
+    Set<TimeRange> rangesToBeRemoved = new HashSet<TimeRange>();
+    if(optionalAttendees.size() == 0){
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: isEmpty

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 29]
```diff
@@ -15,9 +15,152 @@
...
+                    for(TimeRange tr : timeranges){
+                        if(tr.overlaps(event.getWhen())){
+                            rangesToBeRemoved.add(tr);
+                        }
+                        ++currentindex;
+                    } 
+                }
+            }
+        }
+    }
```
Comment (KillianLeC) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:154]: How does this differentiate between doAttendeesOverlap?

**Review by KillianLeC — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 29 above)
Comment (RossJ24) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:154]: It tries to fit in optional attendees if it can

**Pull request merged**